### DISCUSSION TITLE
LineChart - How to handle large X axis in a timeline with SideTitles (while using interval = 1)?

### DISCUSSION THREAD

**Discussion by buchen** [imaNNeo/fl_chart#1477] [Q&A]
I am using LineChart to plot data against arbitrary time periods.
Sometimes those time periods are very long (say 10 years and more).
Depending on the chose reporting period, I only show selected dates on the Y axis.
Sometimes, the first days of a quarter, or only the years, or only every 3 years, or only every 5 or 10 years.

**My issue: in order to decide which labels to draw, I set the ```interval = 1``` in the [SideTitles](https://github.com/imaNNeo/fl_chart/blob/master/repo_files/documentations/base_chart.md#SideTitles). However, that makes the layouting slow for long reporting periods - there are too many (empty) widgets created**

I think I cannot use the any interval other than 1, because a) the reporting period can start at any time (there is no "starting offset"), and b) quarters or years do not have a fixed length (think about leap years).

What are my options to address this?
* Apparently, in older version there was a ```checkToShowTitle``` method. Are there plans to re-introduce such a method?
* Is it possible to "hook into" a custom renderer only for the Y axis of the LineChart?
* Develop a separate widget that is positioned below the graph?

The chart looks something like this:

<img width="259" alt="Bildschirmfoto 2023-11-03 um 18 43 36" src="https://github.com/imaNNeo/fl_chart/assets/587976/cf16a582-5a5c-44eb-a120-80e3c3cb172b">

Here are some code fragments as of today:
```dart
AxisTitles(
      sideTitles: SideTitles(
        showTitles: true,
        reservedSize: 20,
        interval: 1,
        getTitlesWidget: bottomTitleWidgets,
      ),
    );
```

```dart
  Widget bottomTitleWidgets(double value, TitleMeta meta) {
    final date = LocalDate.fromEpochDay(value.toInt());

    // the showLabel functions determines which to draw: quarters, years, every 3 years, ...
    if (!showLabel(date)) return Container();

    final text = date.toString('dd.MM.yy');
    return SideTitleWidget(
      axisSide: meta.axisSide,
      fitInside: SideTitleFitInsideData.fromTitleMeta(meta),
      child: Text(text, style: const TextStyle(fontSize: 10)),
    );
  }
```

Looking at the profiling information, drawing the data series with that many data points is not the problem. The layouting of the axis widgets appears to be the problem. If I remove the axis titles (or set a sufficiently high interval), the the performance is acceptable.

I appreciate any hints.

**Comment by imaNNeo** [comment]
Hi, what if you pass an empty `Container()` or `SizedBox()` widget for when you don't want to show anything? Just like what `checkToShowTitle()` supposed to do.

**Comment by buchen** [comment]
> what if you pass an empty ```Container()``` or ```SizedBox()``` widget for when you don't want to show anything?

That is what I do - I return a Container. Returning a SizedBox does not seem to make a difference.

```
if (!showLabel(date)) return Container();
```

I am no expert here. Looking at the CPU profiler, it shows that a lot of time is spent in performLayout, and visitChildrenForSemantics. All this is gone if I pick an interval of say 50. But, as I pointed out above, fixed intervals do not work for a timeline as years do not have a fixed number of days.

**Comment by buchen** [comment]
BTW, I ended up writing a custom widget for my data axis in order to have full control as to when draw labels on the X axis (sometimes yearly, quarterly, monthly).
